ethereum/contracts/pyth/PythGovernance.sol

```diff
@@ -0,0 +1,118 @@
+
+        require(vm.sequence > lastExecutedGovernanceSequence(), "VAA is older than the last executed governance VAA");
+
+        setLastExecutedGovernanceSequence(vm.sequence);
+
+        return vm;
+    }
+
+    function executeGovernanceInstruction(bytes calldata encodedVM) public {
+        IWormhole.VM memory vm = verifyGovernanceVM(encodedVM);
+
+        GovernanceInstruction memory gi = parseGovernanceInstruction(vm.payload);
+
+        require(gi.targetChainId == chainId() || gi.targetChainId == 0, "Invalid target chain for this governance instruction");
+
+        // We are explicitly checking with number as with enum there might be confusions
+        // about the numbers.
+        if (gi.action == GovernanceAction.UpgradeContract) {
+            require(gi.targetChainId != 0, "Upgrade for all the chains does not exists");
+            upgradeContract(gi.payload);
+        } else if (gi.action == GovernanceAction.SetGovernanceDataSource) {
+            setGovernanceDataSource(gi.payload);
+        } else if (gi.action == GovernanceAction.SetDataSources) {
+            setDataSources(gi.payload);
+        } else if (gi.action == GovernanceAction.SetFee) {
+            setFee(gi.payload);
+        } else if (gi.action == GovernanceAction.SetValidPeriod) {
+            setValidPeriod(gi.payload);
+        } else {
+            revert("Invalid governance action");
+        }
+    }
+
+    function upgradeContract(bytes memory encodedPayload) internal {
+        UpgradeContractPayload memory payload = parseUpgradeContractPayload(encodedPayload); 
+        // This contract does not have enough access to execute this, it should be executed on the
+        // upgradable
+        upgradeUpgradableContract(payload);
+    }
+
+    function upgradeUpgradableContract(UpgradeContractPayload memory payload) virtual internal;
+
+    function setGovernanceDataSource(bytes memory encodedPayload) internal {
+        SetGovernanceDataSourcePayload memory payload = parseSetGovernanceDataSourcePayload(encodedPayload);
+
+        PythInternalStructs.DataSource memory oldGovernanceDatSource = governanceDataSource();
+
+        setGovernanceDataSource(payload.newGovernanceDataSource);
+        setLastExecutedGovernanceSequence(payload.initialSequence);
+
+        emit GovernanceDataSourceSet(oldGovernanceDatSource, governanceDataSource(), lastExecutedGovernanceSequence());
+    }
+
+    function setDataSources(bytes memory encodedPayload) internal {
+        SetDataSourcesPayload memory payload = parseSetDataSourcesPayload(encodedPayload);
+
+        PythInternalStructs.DataSource[] memory oldDataSources = validDataSources();
+
+        for (uint i = 0; i < oldDataSources.length; i += 1) {
+            _state.isValidDataSource[keccak256(abi.encodePacked(oldDataSources[i].chainId, oldDataSources[i].emitterAddress))] = false;
```

Comment:
nitpick: extract the code to hash the data source (the `keccak256(...)` bit) into a helper function to reduce code reuse